hpx/parallel/algorithms/inner_product.hpp

```diff
@@ -45,12 +82,60 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
             template <typename ExPolicy, typename InIter1, typename InIter2,
                 typename T_, typename Op1, typename Op2>
             static T
-            sequential(ExPolicy, InIter1 first1, InIter1 last1, InIter2 first2,
-                T_ && init, Op1 && op1, Op2 && op2)
+            sequential(ExPolicy && policy, InIter1 first1, InIter1 last1,
+                InIter2 first2, T_ init, Op1 && op1, Op2 && op2)
             {
-                return std::inner_product(
-                    first1, last1, first2, std::forward<T_>(init),
-                    std::forward<Op1>(op1), std::forward<Op2>(op2));
+                if (first1 == last1)
+                    return std::move(init);
+
+                // check whether we should apply vectorization
+                if (!util::loop_optimization(policy, first1, last1))
+                {
+                    util::loop2(
+                        parallel::v1::seq, first1, last1, first2,
+                        inner_product_partition<
+                            parallel::v1::sequential_execution_policy,
+                            Op1, Op2, T
+                        >{parallel::v1::seq, op1, op2, init});
+
+                    return init;
+                }
+
+                // loop_step properly advances the iterators
+                auto part_sum = util::loop_step(policy,
+                    inner_product_indirect<Op2>{op2}, first1, first2);
```

Comment:
> Isn't this missing to invoke op1?

No, this calculates the first intermediate value which is then used as the `init` for the subsequent operation.

> What about out of bounds check?

That's done [here](https://github.com/STEllAR-GROUP/hpx/blob/datapar/hpx/parallel/algorithms/inner_product.hpp#L88) and [here](https://github.com/STEllAR-GROUP/hpx/blob/datapar/hpx/parallel/datapar/loop.hpp#L109)